### DISCUSSION TITLE
Maplibre Waypoint Arrival

### DISCUSSION THREAD

**Discussion by BurnerMaps** [maplibre/maplibre-navigation-android#118] [Q&A]
Hi there,

I'm working on an application that uses Graphhopper and Maplibre for turn by turn navigation on routes that have many waypoints. 

It is important that I know what waypoint I'm crossing and when it happens. 

In terms of when it happens, up to this point, the best method I've found is to constantly monitor routeprogress and detect when the legs change, but this feels like a workaround. Mapbox has an [arrival event](https://docs.mapbox.com/android/navigation/guides/ui-components/arrival-detection/), I'm wondering if there's similar functionality that I've missed in Maplibre.

In terms of what waypoint I'm crossing, I've been keeping an ordered list in the background, but obviously this is prone to error as I must keep the list of waypoints given to graphhopper and this list synced at all times. Currently I see "Waypoint X" in the banner instructions when approaching a waypoint, I haven't been able to find a way to modify this but if I could, I would feed an id list and read the banner instruction as I cross a waypoint. That may be more of a Graphhopper question though.

I'm using release 3.0.0 of maplibre-navigation-android

Thanks for your help

**Comment by Fabi755** [comment]
> In terms of when it happens, up to this point, the best method I've found is to constantly monitor routeprogress and detect when the legs change, but this feels like a workaround. Mapbox has an arrival event, I'm wondering if there's similar functionality that I've missed in Maplibre.

You use the correct way. At moment the project don't have the callbacks that Mapbox provides. And if we build something, we will internal use the same logic (watch route progress).

Unfortunately I have no experience or answer to your second Graphhopper question.

**Reply by BurnerMaps** [reply]
Thank you!

**Comment by boldtrn** [comment]
To be honest I don't really understand the GraphHopper question. Would you like to have other info in the Banner? I guess that could be done on the GraphHopper side, but I don't really understand the use case yet.

**Reply by BurnerMaps** [reply]
Lets say I have 3 locations to visit and each location has a package to collect. Let's also say these packages have ids [1A,C5,D2].

My understanding is that Graphhopper will give me back a 3 leg journey with each ending at a package location snapped to the street. My question is, is there some way for me to know what package id I am arriving at without keeping a continuous track of the order initially given to Graphhopper. I.e. could I map the ids to the legs so that I could pull the id of a package from the route progress listener for example.

I know that I could track the order and use that, but if possible I'd like to always know what ID the current leg of the journey is targeting.

Thanks

**Reply by boldtrn** [reply]
Ah, you let them optimize the journey? I guess I would match by coordinates. I don't think there is a direct way right now.
